File: example01databinding/src/main/java/foo/bar/example/asafdatabinding/feature/wallet/Wallet.java
package foo.bar.example.asafdatabinding.feature.wallet;


import co.early.fore.core.Affirm;
import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.observer.ObservableImp;

/**
 *
 */
public class Wallet extends ObservableImp{

    private static final String TAG = Wallet.class.getSimpleName();

    private final Logger logger;
    public final int totalDollarsAvailable = 10;
    private int mobileWalletDollars = 0;


    public Wallet(Logger logger) {
        super(WorkMode.SYNCHRONOUS, logger);
        this.logger = Affirm.notNull(logger);
    }


    public void increaseMobileWallet() {
        if (canIncrease()) {
            mobileWalletDollars++;
            logger.i(TAG, "Increasing mobile wallet to:" + mobileWalletDollars);
            notifyObservers();
        }
    }

    public void decreaseMobileWallet(){
        if (canDecrease()) {
            mobileWalletDollars--;
            logger.i(TAG, "Decreasing mobile wallet to:" + mobileWalletDollars);
            notifyObservers();
        }
    }

    public int getMobileWalletAmount(){
        return mobileWalletDollars;
    }

    public int getSavingsWalletAmount(){
        return totalDollarsAvailable - mobileWalletDollars;
    }

    public boolean canIncrease(){
      return mobileWalletDollars<totalDollarsAvailable;
    }

    public boolean canDecrease(){
        return mobileWalletDollars>0;
    }

}

File: example-kt-01reactiveui/src/main/java/foo/bar/example/forereactiveui/feature/wallet/Wallet.kt
package foo.bar.example.forereactiveui.feature.wallet

import co.early.fore.kt.core.logging.Logger
import co.early.fore.core.observer.Observable
import co.early.fore.kt.core.observer.ObservableImp

/**
 * Copyright © 2015-2020 early.co. All rights reserved.
 */
class Wallet(private val logger: Logger) : Observable by ObservableImp() {

    val totalDollarsAvailable = 10
    var mobileWalletAmount = 0
        private set

    fun increaseMobileWallet() {
        if (canIncrease()) {
            mobileWalletAmount++
            logger.i("Increasing mobile wallet to:$mobileWalletAmount")
            notifyObservers() // our state has changed, so we need to let our observers know
        }
    }

    fun decreaseMobileWallet() {
        if (canDecrease()) {
            mobileWalletAmount--
            logger.i("Decreasing mobile wallet to:$mobileWalletAmount")
            notifyObservers() // our state has changed, so we need to let our observers know
        }
    }

    val savingsWalletAmount: Int
        get() = totalDollarsAvailable - mobileWalletAmount

    fun canIncrease(): Boolean = mobileWalletAmount < totalDollarsAvailable

    fun canDecrease(): Boolean = mobileWalletAmount > 0
}
